{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### IGV TE insertions snapshots\n",
    "\n",
    "We want to have an outmated way of generating IGV screenshots for the TE insertions. Then, we will use the screenshots for visual confirmation of TE insertions and transduction calls predicted by TraFic. \n",
    "\n",
    "**Steps**:\n",
    "1. Script to produce a bed file with the TE insertion coordinates from TraFic.\n",
    "2. Script to generate IGV batch file for running IGV\n",
    "3. Execute IGV with the batch file to produce the screenshots"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### 1. Script to produce a bed file with the TE insertion and transduced region coordinates\n",
    "\n",
    "This script will take as input the TraFiC somatic TE insertions. Format:\n",
    "\n",
    "1. Project Id\n",
    "2. Donor Id\n",
    "3. Tumour Id\n",
    "4. Chr (insertion)\n",
    "5. End + cluster (insertion)\n",
    "6. Beginning - cluster (insertion)\n",
    "7. Family TE (L1, ALU, OTHER (SVA), ERVK)\n",
    "8. Gene affected\n",
    "9. Insertion type. Tdo0(solo), Td1 (partnered transduccion), Td2 (Orphan transduction)\n",
    "10. Chr (source element)\n",
    "11. Pos1 (source element)\n",
    "12. Pos2 (source element)\n",
    "13. Orientation (+, - or putative if strand was not determined) (source element)\n",
    "14. Pos1 (transduced region. In Td1 only one, since the beginning will be the end of the source element. Position 1 or 2 depending on the strand of the source element)\n",
    "15. Pos2 (transduced region. In Td1 only one. Same as in Pos1)\n",
    "16. Distance to source element (- if putative)\n",
    "17. Size transduced region\n",
    "18. Number of reads supporting + cluster\n",
    "19. Number of reads supporting - cluster\n",
    "\n",
    "For a Td0, Td1 and Td2 we need to produce a bed entry containing the TE insertion coordinates. \n",
    "\n",
    "Bed format:\n",
    "1. chr - 4th field in somatic \n",
    "2. chrStart - 5th field in somatic \n",
    "3. chrEnd - 6th field in somatic\n",
    "4. name - Composed by *insertionBkp-7th-9th:4th_5th_6th*\n",
    "\n",
    "Additionally, for the Td1 and Td2 we need to produce a bed entry containing the transduced region coordinates. These coordinates will vary depending on the situation:\n",
    "\n",
    "* Td1 and Td 2 (Source element strand determined)\n",
    "    * Plus Strand\n",
    "        * chr - 10th field in somatic\n",
    "        * beg - Pos1 source element (11th)\n",
    "        * end - Pos2 transduced region (15th)\n",
    "        * name - Composed by *trasducedRegion-7th-9th:10th_11th_15th*\n",
    "        \n",
    "    * Minus Strand\n",
    "        * chr - 10th field in somatic\n",
    "        * beg - Pos1 transduced region (14th)\n",
    "        * end - Pos2 source element (12th)\n",
    "        * name - Composed by *trasducedRegion-7th-9th:10th_14th_12th*\n",
    "    \n",
    "* Td1 or Td0 (Strand do not determined, putative)\n",
    "    * chr - 10th field in somatic\n",
    "    * beg - Pos1 transduced region (14th) - margin (5.000bp default)\n",
    "    * end - Pos2 transduced region (15th) + margin (5.000bp default)\n",
    "    * name - Composed by *trasducedRegion-7th-9th:10th_11th_12th*\n",
    "    "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "# User´s input \n",
    "inputFile = \"/Users/brodriguez/Research/Projects/Pancancer/Somatic/Analysis/Igv_snapshots/Test2/4283773c-c04b-4f33-a398-f3566f6d293a_somaticInsertions.txt\"\n",
    "name = \"4283773c-c04b-4f33-a398-f3566f6d293a\" \n",
    "margin = 5000\n",
    "outDir = \"/Users/brodriguez/Research/Projects/Pancancer/Somatic/Analysis/Igv_snapshots/Test2\"\n",
    "\n",
    "## Open input and output files\n",
    "somaticInsertions=open(inputFile, 'r')\n",
    "\n",
    "fileName = name + \".bed\"\n",
    "outFilePath = outDir + \"/\" + fileName\n",
    "outFile = open( outFilePath, \"w\" )\n",
    "\n",
    "# Read input file line by line\n",
    "for line in somaticInsertions:\n",
    "    line = line.rstrip('\\n')\n",
    "    line = line.split(\"\\t\")\n",
    "    projectId = line[0]\n",
    "    donorId = line[1]\n",
    "    tumourId = line[2]\n",
    "    insertChr = line[3]\n",
    "    insertBeg = line[4]\n",
    "    insertEnd = line[5]\n",
    "    family = line[6]\n",
    "    gene = line[7]\n",
    "    insertType = line[8]\n",
    "    sourceChr = line[9]\n",
    "    sourcePos1 = line[10]\n",
    "    sourcePos2 = line[11]\n",
    "    strand = line[12]\n",
    "    transducedPos1 = line[13]\n",
    "    transducedPos2 = line[14]\n",
    "    dist2source = line[15]\n",
    "    size = line[16]\n",
    "    nbReadsPlus = line[17]\n",
    "    nbReadsMinus = line[18]\n",
    "\n",
    "    # Discard donors without any transposable element (TE) insertion \n",
    "    if family != \"empty\":  \n",
    "        \n",
    "        # Report TE insertion coordinates in bed format\n",
    "        featureName = \"insertionBkp\" + \"-\" + family + \"-\" + insertType + \":\" + insertChr + \"_\" + insertBeg + \"_\" + insertEnd  \n",
    "        row = insertChr + \"\\t\" + insertBeg + \"\\t\" + insertEnd + \"\\t\" + featureName + \"\\n\"\n",
    "        outFile.write(row)\n",
    "        \n",
    "        # Additionally, report transduction coordinates if it is a Td1 \n",
    "        # (partnered transduccion) or a Td2 (Orphan transduction)\n",
    "        # Coordinates will depend on the orientation of the TE\n",
    "        if insertType == \"td1\" or insertType == \"td2\":\n",
    "            \n",
    "            # A) TE orientation unknown:   \n",
    "            # ---source?---  ---TransducedRegion---  ---source?---\n",
    "            if strand == \"putative\":\n",
    "                beg = int(transducedPos1) - margin\n",
    "                end = int(transducedPos2) + margin\n",
    "            \n",
    "            # B) TE in positive strand\n",
    "            # ---source---  ---TransducedRegion---  \n",
    "            elif strand == \"plus\":\n",
    "                beg = sourcePos1\n",
    "                end = transducedPos2\n",
    "            \n",
    "            # C) TE in negative strand\n",
    "            # ---TransducedRegion---  ---source---\n",
    "            else:\n",
    "                beg = transducedPos1\n",
    "                end = sourcePos2\n",
    "            \n",
    "            # Report the coordinates\n",
    "            featureName = \"transducedRegion\" + \"-\" + family + \"-\" + insertType + \":\" + insertChr + \"_\" + insertBeg + \"_\" + insertEnd \n",
    "            row = sourceChr + \"\\t\" + str(beg) + \"\\t\" + str(end) + \"\\t\" + featureName + \"\\n\"\n",
    "            outFile.write(row)\n",
    "        \n",
    "# Close output and end\n",
    "outFile.close()\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### 2. Script to generate an IGV batch file for running IGV\n",
    "\n",
    "This script will take as input a bed file with the Trafic insertion coordinates and a bam file plus other optional parameters: \n",
    "\n",
    "* **Mandatory**\n",
    "    * Input bed\n",
    "    \n",
    "        1. chr \n",
    "        2. chrStart \n",
    "        3. chrEnd \n",
    "        4. name \n",
    "    \n",
    "    * Input bam\n",
    "\n",
    "\n",
    "* **Optional**\n",
    "    * Genome Version. Default: hg19\n",
    "    * Track in bed format. E.g: repeat masker repeats. \n",
    "    * Margin. Number of bases around the range defined in the bed entry to be included in the snapshot. \n",
    "    * maxPanelHeight. Sets the number of vertical pixels (height) of each panel to include in image. Default: 1000\n",
    "    * Format. Image file format, which must be png, jpg, or svg. Default: png\n",
    "    * OutDir. Default: current working directory. \n",
    "\n",
    "\n",
    "It will produce as output an IGV batch file like this one:\n",
    "\n",
    "```\n",
    "new\n",
    "genome hg19\n",
    "load /Users/brodriguez/Research/Projects/Pancancer/Igv_snapshots/db2a1df8-487d-4dad-a347-5c5ed539d5ad.bam\n",
    "load /Users/brodriguez/Research/Projects/Pancancer/Igv_snapshots/repeatmasker_data_from_ucsc_090112.sorted.bed\n",
    "snapshotDirectory /Users/brodriguez/Research/Projects/Pancancer/Igv_snapshots \n",
    "goto chrX:19532036-19534134\n",
    "snapshot \n",
    "goto chrX:39964716-39966767\n",
    "snapshot \n",
    "goto chr20:38234854-38236804\n",
    "snapshot\n",
    "exit\n",
    "```\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "# User´s input \n",
    "inputBed = \"/Users/brodriguez/Research/Projects/Pancancer/Somatic/Analysis/Igv_snapshots/Test2/4283773c-c04b-4f33-a398-f3566f6d293a.bed\"\n",
    "inputBam = \"/Users/brodriguez/Research/Projects/Pancancer/Somatic/Analysis/Igv_snapshots/Test2/4283773c-c04b-4f33-a398-f3566f6d293a.bam\"\n",
    "fileName = \"igv_batch.txt\"\n",
    "genome = \"hg19\"\n",
    "track = \"/Users/brodriguez/Research/Projects/Pancancer/Somatic/Analysis/Igv_snapshots/Test1/repeatmasker_data_from_ucsc_090112.sorted.bed\"\n",
    "margin = \"1000\"\n",
    "maxPanelHeight = \"500\"\n",
    "imgFormat = \"png\" \n",
    "outDir = \"/Users/brodriguez/Research/Projects/Pancancer/Somatic/Analysis/Igv_snapshots/Test3\"\n",
    "\n",
    "## Open input and output files\n",
    "bed = open(inputBed, 'r')\n",
    "\n",
    "outFilePath = outDir + \"/\" + fileName\n",
    "outFile = open( outFilePath, \"w\" )\n",
    "\n",
    "## Add basic tasks (create session, load genome, bam, optional track...) to batch file\n",
    "outFile.write(\"%s\\n\" % \"new\")\n",
    "outFile.write(\"%-1s %s\\n\" % (\"genome\", genome))\n",
    "outFile.write(\"%-1s %s\\n\" % (\"load\", inputBam))\n",
    "\n",
    "if track != \"none\":\n",
    "    outFile.write(\"%-1s %s\\n\" % (\"load\", track))    \n",
    "\n",
    "## Iterate over the input bed file and create the tasks to take snapshots\n",
    "# of the regions of interest:\n",
    "for line in bed:\n",
    "    line = line.rstrip('\\n')\n",
    "    line = line.split(\"\\t\")\n",
    "    chrom = line[0]\n",
    "    beg = int(line[1]) - int(margin) \n",
    "    end = int(line[2]) + int(margin)\n",
    "    snapshotPos = \"chr\" + str(chrom) + \":\" + str(beg) + \"-\" + str(end)\n",
    "    snapshotName = line[3] + \".\" +  imgFormat\n",
    "    \n",
    "    # Write tasks to take the snapshots into the batch file \n",
    "    outFile.write(\"%-1s %s\\n\" % (\"goto\", snapshotPos))\n",
    "    outFile.write(\"%-1s %s\\n\" % (\"maxPanelHeight\", maxPanelHeight))\n",
    "    outFile.write(\"%-1s %s\\n\" % (\"snapshot\", snapshotName))\n",
    "\n",
    "# Add exit task\n",
    "outFile.write(\"%s\\n\" % \"exit\")\n",
    "\n",
    "# Close batch file and ends\n",
    "outFile.close()\n",
    "    "
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### 2. Execute IGV with the batch file to produce the screenshots\n",
    "\n",
    "### 2.1 Test 1 files\n",
    "\n",
    "I will execute the IGV with the batch file:\n",
    "\n",
    "```\n",
    "igv=/Users/brodriguez/Research/Apps/IGV/2.3.72/igv.sh\n",
    "batchFile=/Users/brodriguez/Research/Projects/Pancancer/Somatic/Analysis/Igv_snapshots/Test1/igv_batch.txt\n",
    "outDir=/Users/brodriguez/Research/Projects/Pancancer/Somatic/Analysis/Igv_snapshots/Test1/Screenshots/\n",
    "cd $outDir\n",
    "\n",
    "time $igv -g hg19 -b $batchFile\n",
    "real: 41m3.118s\n",
    "user: 2m9.247s\n",
    "sys: 0m10.458s\n",
    "```\n",
    "\n",
    "Images look good. It took ~41' to generate the 532 imagess corresponding to the 532 insertions in one donor\n",
    "\n",
    "Generate the images for the transduced regions for the 2 td2 insertions we have:\n",
    "\n",
    "```\n",
    "igv=/Users/brodriguez/Research/Apps/IGV/2.3.72/igv.sh\n",
    "batchFile=/Users/brodriguez/Research/Projects/Pancancer/Somatic/Analysis/Igv_snapshots/Test1/igv_batch_transducedRegions.txt\n",
    "outDir=/Users/brodriguez/Research/Projects/Pancancer/Somatic/Analysis/Igv_snapshots/Test1/Screenshots/\n",
    "cd $outDir\n",
    "\n",
    "time $igv -g hg19 -b $batchFile\n",
    "```\n",
    "\n",
    "The images looks good.\n",
    "\n",
    "### 2.2 Test 2 files\n",
    "As test1 donor did not contain many transductions I asked jose for the minibam from another donor. \n",
    "\n",
    "He sent me one with 19 td1 and 61 td2. I will execute the IGV on the already generated batch file:\n",
    "\n",
    "```\n",
    "igv=/Users/brodriguez/Research/Apps/IGV/2.3.72/igv.sh\n",
    "batchFile=/Users/brodriguez/Research/Projects/Pancancer/Somatic/Analysis/Igv_snapshots/Test2/igv_batch.txt\n",
    "outDir=/Users/brodriguez/Research/Projects/Pancancer/Somatic/Analysis/Igv_snapshots/Test2/Screenshots/\n",
    "cd $outDir\n",
    "\n",
    "time $igv -g hg19 -b $batchFile\n",
    "real\t52m58.189s\n",
    "user\t2m44.327s\n",
    "sys\t0m14.706s\n",
    "```\n",
    "\n",
    "#### Pick an example of each type to verify the image is fine: \n",
    "* Td0 (solo-L1)\n",
    "\n",
    "```\n",
    "grep 'td0' 4283773c-c04b-4f33-a398-f3566f6d293a_somaticInsertions.txt | head -1\n",
    "ESAD-UK\t4283773c-c04b-4f33-a398-f3566f6d293a\tOCCAMS-AH-091\tX\t5593350\t5593301\tL1\tRP11-733O18.1\ttd0\t-\t-\t-\t-\t-\t-\t-\t7\n",
    "\n",
    "cd /Users/brodriguez/Research/Projects/Pancancer/Somatic/Analysis/Igv_snapshots/Test2/Screenshots\n",
    "open insertionBkp-L1-td0:X_5593350_5593301.png\n",
    "Ok, it looks good, multicolored positive and negative clusters\n",
    "```\n",
    "    \n",
    "* Td1 (partnered-transduccion)\n",
    "    * putative\n",
    "    \n",
    "    ```\n",
    "    grep 'td1' 4283773c-c04b-4f33-a398-f3566f6d293a_somaticInsertions.txt | grep 'putative' | head -1\n",
    "    ESAD-UK\t4283773c-c04b-4f33-a398-f3566f6d293a\tOCCAMS-AH-091\t1\t37673028\t37672993\tL1\t-\ttd1\t4\t117340987\t117341153\tputative 117341050\t117341217\t-\t-\t5\t10\t5\n",
    "    \n",
    "    cd /Users/brodriguez/Research/Projects/Pancancer/Somatic/Analysis/Igv_snapshots/Test2/Screenshots\n",
    "    \n",
    "    open insertionBkp-L1-td1:1_37673028_37672993.png\n",
    "    Ok, looks good, single colored positive and multicolored negative cluster\n",
    "    \n",
    "    open transducedRegion-L1-td1:1_37673028_37672993.png\n",
    "    Ok, looks good, multicolored cluster with a subset of reads with mates on the same region (blue reads). Putative source L1PA up or down stream\n",
    "    ```\n",
    "  \n",
    "    * plus\n",
    "    ```\n",
    "    grep 'td1' 4283773c-c04b-4f33-a398-f3566f6d293a_somaticInsertions.txt | grep 'plus' | head -1\n",
    "    ESAD-UK\t4283773c-c04b-4f33-a398-f3566f6d293a\tOCCAMS-AH-091\tX\t146016754\t146016705\tL1\t-\ttd1\t22\t29059271\t29065303\tplus\t29065497\t194\t-\t19\t18\t19\n",
    "    \n",
    "    open insertionBkp-L1-td1:X_146016754_146016705.png\n",
    "    Ok, weird.. two colored positive and multicolored negative cluster. Ask José\n",
    "    \n",
    "    open transducedRegion-L1-td1:X_146016754_146016705.png\n",
    "    Ok, looks good, cluster composed by set of reads of the same colours (they do correspond to different transduction events). Source L1H upstream\n",
    "    ```\n",
    "    \n",
    "    * minus\n",
    "    ```\n",
    "    grep 'td1' 4283773c-c04b-4f33-a398-f3566f6d293a_somaticInsertions.txt | grep 'minus' | head -1\n",
    "    ESAD-UK\t4283773c-c04b-4f33-a398-f3566f6d293a\tOCCAMS-AH-091\t7\t145833301\t145833294\tL1\tCNTNAP2\ttd1\t4\t137214649\t137220701\tminus\t137213866\t-\t783\t-\t7\t6\t7\n",
    "    \n",
    "    open insertionBkp-L1-td1:7_145833301_145833294.png\n",
    "    Ok, multicolored positive and singlecolored negative cluster\n",
    "    \n",
    "    open transducedRegion-L1-td1:7_145833301_145833294.png\n",
    "    Ok, looks good, cluster composed by set of reads of the same colours (they do correspond to different transduction events). Source L1H downstream\n",
    "    ```\n",
    "    \n",
    "* Td2 (orphan-transduccion)\n",
    "    * putative\n",
    "    ```\n",
    "    grep 'td2' 4283773c-c04b-4f33-a398-f3566f6d293a_somaticInsertions.txt | grep 'putative' | head -1\n",
    "    ESAD-UK\t4283773c-c04b-4f33-a398-f3566f6d293a\tOCCAMS-AH-091\tX\t58154067\t58154025\tL1\t-\ttd2\t6\t13190686\t13190980\tputative 13191016\t13191505\t-\t489\t9\t10\t9\n",
    "    \n",
    "    cd /Users/brodriguez/Research/Projects/Pancancer/Somatic/Analysis/Igv_snapshots/Test2/Screenshots\n",
    "    \n",
    "    open insertionBkp-L1-td2:X_58154067_58154025.png\n",
    "    Ok, looks good, single colored positive and negative clusters\n",
    "    \n",
    "    open transducedRegion-L1-td2:X_58154067_58154025.png\n",
    "    Ok, looks good. Transduction cluster + somatic source element cluster\n",
    "    ```\n",
    "  \n",
    "    * plus\n",
    "    ```\n",
    "    grep 'td2' 4283773c-c04b-4f33-a398-f3566f6d293a_somaticInsertions.txt | grep 'plus' | head -1\n",
    "    ESAD-UK\t4283773c-c04b-4f33-a398-f3566f6d293a\tOCCAMS-AH-091\t2\t36358155\t36358126\tL1\t-\ttd2\t20\t23406745\t23412777\tplus\t23412963\t23413658\t881\t695\t4\t4\t4\n",
    "    \n",
    "    cd /Users/brodriguez/Research/Projects/Pancancer/Somatic/Analysis/Igv_snapshots/Test2/Screenshots\n",
    "    \n",
    "    open insertionBkp-L1-td2:2_36358155_36358126.png\n",
    "    Ok, looks good, single colored positive and negative clusters\n",
    "    \n",
    "    open transducedRegion-L1-td2:2_36358155_36358126.png\n",
    "    Ok, looks good, cluster composed by set of reads of the same colours (they do correspond to different transduction events). Source L1H upstream\n",
    "    ```\n",
    "    \n",
    "    * minus\n",
    "    ```\n",
    "    grep 'td2' 4283773c-c04b-4f33-a398-f3566f6d293a_somaticInsertions.txt | grep 'minus' | head -1\n",
    "    ESAD-UK\t4283773c-c04b-4f33-a398-f3566f6d293a\tOCCAMS-AH-091\tX\t9823549\t9823486\tL1\tSHROOM2\ttd2\t4\t137214649\t137220701\tminus\t137213864\t137214653\t785\t789\t7\t4\t7\n",
    "    \n",
    "    open insertionBkp-L1-td2:X_9823549_9823486.png\n",
    "    Ok, looks good, single colored positive and negative clusters\n",
    "    \n",
    "    open transducedRegion-L1-td2:X_9823549_9823486.png\n",
    "    Ok, looks good, cluster composed by set of reads of the same colours (they do correspond to different transduction events). Source L1H downstream\n",
    "    ```\n",
    "\n",
    "\n",
    "**Ok, as everything works fine I will put the code into two scripts**\n",
    "\n",
    "\n",
    "\n"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 2",
   "language": "python",
   "name": "python2"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 2
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython2",
   "version": "2.7.11"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 0
}
